HB/common/phant-abbreviation.elpi

```diff
@@ -1,73 +1,143 @@
+  assert-good-coverage! MLSortedRevFiltered CNL,
+
+  RMClauses => w-params.then MLwP real-arg real-arg
+    (private.mk-phant-term.classes EtaF CNL) PhBody,
+].
+
+% A la carte API: start with a term and wrap it up
+
+% A term with no phantom arguments
+pred init i:term, o:phant-term.
+init T (private.phant-term [] T).
+
+% [real-arg N T Ph Ph1] Adds a real argument named N of type T around Ph
+pred real-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
+real-arg N T F Res :- !, private.phant-fun (private.real N) T F Res.
+
+% [unify-arg Msg X1 X2 Ph Ph1] Adds an argument that will foce the unification
+% of X1 with X2 and print Msg is case of error around Ph
+pred unify-arg i:option term, i:term, i:term, i:phant-term, o:phant-term.
+unify-arg OMsg X1 X2 (private.phant-term AL F) (private.phant-term [private.idfun|AL] UF) :-
+  std.assert-ok! (coq.typecheck X1 T1) "unify-arg: X1 illtyped",
+  std.assert-ok! (coq.typecheck X2 T2) "unify-arg: X2 illtyped",
+  if (OMsg = some M) (Msg = {{lib:hb.some lp:M}}) (Msg = {{lib:hb.nomsg}}),
+  UF = {{fun unif_arbitrary : lib:hb.unify lp:T1 lp:T2 lp:X1 lp:X2 lp:Msg => lp:F}}.
+
+% [implicit-arg N T Ph Ph1] Adds an implicit argument name N of type T areound Ph
+pred implicit-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
+implicit-arg N Ty PF (private.phant-term [private.implicit|AL] (fun N Ty F)) :- !,
+  @pi-decl N Ty t\ PF t = private.phant-term AL (F t).
+
+% [infer-arg N T Ph Ph1] Adds an argument N of type T such that one passes 
+% a value V of type {{ Type }} the corresponding canonical VC of type T is passed
+% for N , eg `fun T (phT : phant T) => Ph`
+pred infer-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
+infer-arg N Ty PF Out :- !,
+  @pi-decl N Ty t\ sigma PhN\
+    PF t = private.phant-term AL (Bo t),
+    coq.string->name {calc ("ph" ^ {coq.name->id N})} PhN,
+    PF' t = private.phant-term [private.infer N|AL] (fun PhN {{ lib:@hb.phant lp:t }} _\ Bo t),
+    implicit-arg N Ty PF' Out.
+
+% TODO: this looks like a hack to remove
+pred append-phant-unify i:phant-term, o:phant-term.
+append-phant-unify (private.phant-term LP T) (private.phant-term LPU T) :-
+  std.append LP [private.idfun] LPU.
+
+/* ------------------------------------------------------------------------- */
+/* ----------------------------- private code ------------------------------ */
+/* ------------------------------------------------------------------------- */
+
+namespace private {
 
 % phant-term is a pair of a list of argument kinds together with a term
 kind phant-term type.
 type phant-term list phant-arg -> term -> phant-term.
 
+% phant-arg encode these three kind of arguments
+% - [x_i] is encoded using [real x_i]
+% - [_]              using [implicit]
+% - [id]             using [idfun]
+% - [Phant x]        using [infer]
+kind phant-arg type.
+type real name -> phant-arg.
+type infer name -> phant-arg.
+type implicit phant-arg.
+type idfun phant-arg.
```

Comment:
The name `idfun` reveals more about the implementation than it should. `unify` is the semantics.